src/Storage/Hydration/EventHydrator.php

```diff
@@ -0,0 +1,167 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Krixon\DomainEvent\Storage\Hydration;
+
+use Krixon\DomainEvent\Event;
+use Krixon\DomainEvent\Storage\Hydration\Exception\HydrationFailed;
+use Krixon\DomainEvent\Storage\Hydration\Resolver\EventClassResolver;
+use ReflectionClass;
+use ReflectionException;
+use ReflectionMethod;
+use ReflectionProperty;
+use function array_merge;
+use function sprintf;
+use function ucfirst;
+
+class EventHydrator
+{
+    private $classResolver;
+
+
+    public function __construct(EventClassResolver $classResolver)
+    {
+        $this->classResolver = $classResolver;
+    }
+
+
+    /**
+     * @param mixed[] $data
+     */
+    public function hydrate(array $data) : Event
+    {
+        if (!isset($data['eventType'])) {
+            throw new HydrationFailed('Cannot hydrate event without an event type.');
+        }
+
+        $class = $this->classResolver->resolve($data['eventType']);
+
+        try {
+            $reflection = new ReflectionClass($class);
+        } catch (ReflectionException $exception) {
+            throw new HydrationFailed(
+                sprintf("Cannot hydrate event, class '%s' could not be found.", $class),
+                0,
+                $exception
+            );
+        }
+
+        $event      = $reflection->newInstanceWithoutConstructor();
+        $properties = $this->getProperties($reflection);
```

Comment:
Minor point, but this is wasted work if we're going to throw below anyway. It could be moved to below that check.

Reply:
Moved